File: 77. Combinations.py
# https://leetcode.com/problems/combinations/

# Given two integers n and k, return all possible combinations of k numbers out of
# the range [1, n].

# You may return the answer in any order.

################################################################################

# dfs

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        self.n = n
        self.k = k
        
        self.ans = []
        self.holder = []
        
        self.dfs(0)
        
        return self.ans
    
    def dfs(self, i):
        if len(self.holder) == self.k:
            self.ans.append(self.holder.copy())
            return
        
        for j in range(i + 1, self.n + 1): # j = i+1...n
            self.holder.append(j)
            self.dfs(j)
            self.holder.pop()            

File: 77. Combinations.cpp
// https://leetcode.com/problems/combinations/

// Given two integers n and k, return all possible combinations of k numbers out of
// the range [1, n].

// You may return the answer in any order.

////////////////////////////////////////////////////////////////////////////////

class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        dfs(0, n, k); return ans;
    }
private:
    vector<int> holder;
    vector<vector<int>> ans;
    void dfs(int i, const int& n, const int& k) {
        if (holder.size() == k) {
            ans.push_back(holder); return;
        }
        for (int j = i + 1; j <= n; ++j) {
            holder.push_back(j);
            dfs(j, n, k);
            holder.pop_back();
        }
        return;
    }
    
};
